File: src/day1/day1.zig
const std = @import("std");
const print = std.debug.print;
const ArrayList = std.ArrayList;

const file_util = @import("../util/file_utils.zig");

pub fn run() !void {
    const allocator = std.heap.page_allocator;
    const lines = try file_util.readFileToLines(allocator, "src/day1/input.txt");
    defer {
        for (lines) |line| {
            allocator.free(line);
        }
    }

    var left_list = ArrayList(u32).init(allocator);
    defer left_list.deinit();
    var right_list = ArrayList(u32).init(allocator);
    defer right_list.deinit();

    for (lines) |line| {
        var it = std.mem.split(u8, line, "   ");
        var i: i8 = 0;
        while (it.next()) |x| {
            const number = try std.fmt.parseInt(u32, x, 10);
            if (i == 0) {
                try left_list.append(number);
            } else {
                try right_list.append(number);
            }
            i += 1;
        }
    }

    // Sort both list in place
    std.mem.sort(u32, left_list.items, {}, std.sort.asc(u32));
    std.mem.sort(u32, right_list.items, {}, std.sort.asc(u32));

    var total_distance: u32 = 0;

    for (left_list.items, right_list.items) |left, right| {
        const diff: u32 = @intCast(@abs(@as(i64, left) - @as(i64, right)));

        print("left {d}, right {d}: diff {d}\n", .{ left, right, diff });

        total_distance += diff;
    }
    print("len {d}\n", .{left_list.items.len});

    print("Total distance: {d}\n", .{total_distance}); //2114951

    var similarity_score: u32 = 0;
    // Slow but this is a small data set
    for (left_list.items) |left| {
        const count: u32 = @intCast(std.mem.count(u32, right_list.items, &[_]u32{left}));
        similarity_score += left * count;
    }
    print("Similarity score {}\n", .{similarity_score});
}

File: src/day1/day1.py
import os
from collections import Counter

from src.util.file_utils import read_file_to_lines

input_file = os.path.dirname(os.path.realpath(__file__)) + "/input.txt"


def run_part1():
    input = read_file_to_lines(input_file)

    left, right = tuple(
        map(
            list,
            zip(
                *[
                    (int(nums[0]), int(nums[1]))
                    for nums in [x.split("   ") for x in input]
                ]
            ),
        ),
    )
    left.sort()
    right.sort()
    result = 0
    for i in range(len(left)):
        result += abs(left[i] - right[i])

    print(result)


run_part1()


def run_part2():
    input = read_file_to_lines(input_file)

    left, right = tuple(
        map(
            list,
            zip(
                *[
                    (int(nums[0]), int(nums[1]))
                    for nums in [x.split("   ") for x in input]
                ]
            ),
        ),
    )
    counts = Counter(right)
    result = 0
    for n in left:
        result += n * counts[n]

    print(result)


run_part2()
